### ISSUE TITLE
Q: Is FFTW3 transpilable to Webassembly (WASM via emscripten)?

### ISSUE THREAD

**Issue by RalphSteinhagen** [FFTW/fftw3#293]
First, thanks a lot for all the effort you put into the de-facto backbone of most numerical tools on the internet! Kudos :+1:

We have some libraries/applications that rely on fftw3 and which we would eventually like to also deploy to the web.
In order to keep the same API, it would be great if fftw3 would transpile to webassembly (e.g. using emscripten).

Did anybody have success transpiling the fftw code to WASM? Are there plans? 

Thanks in advance for a

**Comment by latermarch** [comment]
I build the .a lib via emscripten like this:
 emconfigure ./configure --disable-fortran --enable-static  --disable-single  --enable-threads  --prefix=$EM_PATH CFLAGS="-s WASM=1" CXXFLAGS="-s WASM=1"  LDFLAGS="-s WASM=1"

but when run the web application,it stucked:
![image](https://github.com/FFTW/fftw3/assets/9695216/35f3829f-0a51-4d29-95a0-30538c1b9fdb)

![image](https://github.com/FFTW/fftw3/assets/9695216/8ed638af-e9ff-45f8-b6c9-8a9178504299)
I don't know how to fix it

**Comment by RalphSteinhagen** [comment]
Meanwhile, we found a solution to integrate and compile FFTW3 to WASM. See [here](https://github.com/fair-acc/graph-prototype) and the FFT block for example.

Here is the specific cmake example:
```cmake
# FFTW3 is build 2 times for float and double precisions
SET(FFTW_PREFIX ${CMAKE_CURRENT_BINARY_DIR}/fftw)
SET(FFTW_CFLAGS "${CFLAGS} -w")
if (EMSCRIPTEN)
    SET(FFTW_CONFIG cd ${FFTW_PREFIX}/src/ && emconfigure ./configure --enable-silent-rules --quiet --disable-fortran --prefix=${FFTW_PREFIX}/install)
    SET(FFTW_BUILD emmake make -j CFLAGS=${FFTW_CFLAGS} --silent V=0 && emmake make install --silent V=0 && emmake make clean --silent V=0)
else ()
    SET(FFTW_CONFIG ${FFTW_PREFIX}/src/configure --enable-silent-rules --quiet --disable-fortran --prefix=${FFTW_PREFIX}/install)
    SET(FFTW_BUILD make -j CFLAGS=${FFTW_CFLAGS} --silent V=0 && make install --silent V=0 && make clean --silent V=0)
endif ()
SET(FFTW_INSTALL_COMMAND
        ${FFTW_CONFIG} && ${FFTW_BUILD} &&
        ${FFTW_CONFIG} --enable-float && ${FFTW_BUILD})


include(ExternalProject)
ExternalProject_Add(fftw_ext
        PREFIX ${FFTW_PREFIX}
        SOURCE_DIR ${FFTW_PREFIX}/src
        BINARY_DIR ${FFTW_PREFIX}/build
        INSTALL_DIR ${FFTW_INSTALL_DIR}
        STAMP_DIR ${FFTW_PREFIX}/stamp
        TMP_DIR ${FFTW_PREFIX}/tmp
        DOWNLOAD_DIR ${FFTW_PREFIX}/download
        LOG_DIR ${FFTW_PREFIX}/log
        URL "https://fftw.org/fftw-3.3.10.tar.gz"
        URL_MD5 8ccbf6a5ea78a16dbc3e1306e234cc5c
        CONFIGURE_COMMAND ${FFTW_INSTALL_COMMAND}
        BUILD_COMMAND ""
        INSTALL_COMMAND ""
        LOG_DOWNLOAD ON
)


include_directories(${FFTW_PREFIX}/install/include)


add_library(fftw INTERFACE)
target_link_libraries(fftw INTERFACE fftw3 INTERFACE fftw3f INTERFACE m)
target_include_directories(fftw INTERFACE ${FFTW_PREFIX}/install/include)
target_link_directories(fftw INTERFACE ${FFTW_PREFIX}/install/lib ${FFTW_PREFIX}/install/lib64)
add_dependencies(fftw fftw_ext)
```

or a more basic bash-style version:
```bash
#/bin/bash

if [ -z ${EM_PATH+x} ]; then 
        echo "EM_PATH is unset"; 
        export EM_PATH="$(PWD)/test"
else 
        echo "EM_PATH is set to '$EM_PATH'"; 
fi

wget "http://fftw.org/fftw-3.3.10.tar.gz"
tar -xf fftw-3.3.6-pl2.tar.gz
cd fftw-3.3.6-pl2
emconfigure ./configure --disable-fortran --enable-float --prefix=$EM_PATH CFLAGS="-s WASM=1" CXXFLAGS="-s WASM=1"  LDFLAGS="-s WASM=1"  #--with-combined-threads
emmake make -j
emmake make install
```

Certainly! Below is a revised version of your comment:

Note: The GPLv2 license of FFTW, without linking/WebAssembly (WASM) exceptions, makes its use in commercial settings challenging. While this is fine for us as a public organization, it's a hurdle for some industry and governmental library co-users. As a solution, we're developing a more straightforward, albeit possibly slower, alternative using `std::simd`. This approach aims to minimize maintenance, reduce external dependencies, ease cross-compilation to WASM, and encourage a more public adoption of our prototype library, envisioned as a stepping stone towards 'GNU Radio 4.0'.

**Comment by mildsunrise** [comment]
@RalphSteinhagen do you have some rough idea of how well it performs when transpiled to WASM vs natively? It will obviously vary a bit by computer and WASM engine, but I'm hoping we can get to a 2x factor (e.g. 50% the speed of native), FTTW or not

**Comment by RalphSteinhagen** [comment]
Hey @mildsunrise, yes, we have a 'rough' idea and wrote a [benchmark](https://github.com/fair-acc/gnuradio4/blob/main/algorithm/benchmarks/bm_fft.cpp) since the FFT is crucial for our type of signal processing (GNU Radio 4.0). Have a look at the `ops/s` for typical sizes of N that we are using (sorry for the shifted columns).

## **TLDR**: 
  * WebAssembly SIMD support for FFTs is still rather poor (as of Q2-2025, being worked upon).
  * **WASM**: 
     FFTW (46M ops/s), 'pretty fast FFT' (54M ops/s), and about as  fast as a basic textbook implementation (~50M ops/s): 
  * **Native**: 
     FFTw (2.2G ops/s) is ~2x slower than 'pfFFT' (4.4G Ops/s), and only 2.7x faster than the textbook impl (0.8G ops/s).

### **WASM (Release build, node22)**:

<img width="1020" height="985" alt="Image" src="https://github.com/user-attachments/assets/b7896187-b969-47d3-b36b-e09de2120b74" />

### **Native (Release build, gcc15)**:

<img width="1468" height="985" alt="Image" src="https://github.com/user-attachments/assets/ab338aa1-09d6-4e65-842a-6a080caea256" />

'FFT' is our vanilla textbook FFT, and  'FFTpf' is the 'pretty fast FFT' (we'll be switching over to that one). Have a go at it in case you need different FFT sizes, larger sample sizes (these are configurable/kept very short to get a quick overview), or have suggestions on how to further improve on that. PRs are always welcome. 😄 

<sub><sub>
N.B. We are working on a `std::simd`-fication of the FFT, which should make the porting/vectorisation to WASM/Emscripten a bit easier in the future (if you are using C++, that is).</sub></sub>

**Comment by mildsunrise** [comment]
Wow, that's a very detailed answer, thank you so much!

Good to know FFTW performs more or less like a textbook implementation when compiled to WASM, although a bit depressing :_)
For the record, I ran your benchmark on my computer and FFTW and FFT produce very similar results to my [toy WASM FFT](https://gist.github.com/mildsunrise/d5990a0651009b8c96460c53a3493e03) (literally the first FFT I've written in my life).

Thanks for the pointer to FFTpf, it is indeed the fastest performing WASM implementation I've seen to date...

Saw some of the `std::simd` code, nice :) I might give SIMD a shot now that I know a bit more about WASM and FFTs.

> WebAssembly SIMD support for FFTs is still rather poor (as of Q2-2025, being worked upon).

Is this meant to say that the SIMD instructions currently provided by WebAssembly aren't well suited for FFTs, or that the current FFT WASM implementations don't use them very well?

**Comment by RalphSteinhagen** [comment]
> Is this meant to say that the SIMD instructions currently provided by WebAssembly aren't well suited for FFTs, or that the current FFT WASM implementations don't use them very well?

@mildsunrise the latter. The WASM SIMD extension is well specified. However, one needs to hard-code these, and thus this isn't super portable/integrated into our FFT and other algorithm implementations. 

We are working (albeit at a slow pace) on moving more and more code to using the official `std::simd<T>` extensions that internally map to the platform-specific instructions but play nice when being used on x86, ARM, RISC, ...

**Comment by mildsunrise** [comment]
glad to hear that! I just wrote a WASM SIMD implementation and [a browser benchmark](https://mildsunrise.github.io/web-fft-benchmark/) and in my i7-8665U I'm getting a peak of 12.8 Gflops. Note I'm using FFTw's convention where I multiply by 5N log<sub>2</sub>(N), not N ln(N) as you do, so there's a conversion factor of 5/ln(2) ≈ 7.2. So this would correspond to 1.77 Gops/s in your benchmark. This looks much more acceptable to me, and it can probably be improved further (again, I don't have experience writing FFTs or WASM)

For reference, on this machine your benchmark reports a peak of 466 Mops/s for FFTw on WASM and 600 Mops/s for FFTw native with gcc 14. But benchfft reports 40 Gflops = 5.5 Gops/s in those conditions. Maybe I'm not compiling gnuradio or your benchmark correctly...

**Comment by RalphSteinhagen** [comment]
@mildsunrise took a bit of time, but we are having a header-only liberally-licensed `SimdFFT<T, ...>` implementation now. The simd acceleration needs the Emscripten `-msimd128` or `-mrelaxed-simd`. For details see [here](https://github.com/fair-acc/gnuradio4/pull/671).

Even at the expense of embarrassing myself wrt my inaptitude of writing high-performance SIMD code for WASM, I'd love to know how our FFT rendition performs in your specific WASM example code. 

In any case, your feedback would be much appreciated.

**Comment by mildsunrise** [comment]
oooh, nice! I'll try to include it in the benchmark, it isn't the most portable thing I've ever seen but it seems pretty doable, just vir-simd and a couple gnuradio headers... sadly the benchmark only tests complex FFTs for now, I still haven't gotten around to adding a real FFT mode

**Comment by mildsunrise** [comment]
okay, done! it returns correct results, and is now live in the benchmark. you can see the Makefile and the supporting C and JS wrapper code [here](https://github.com/mildsunrise/web-fft-benchmark/commit/650d245e7f6537272f7160170991ffcc9712e20a). in particular note this is using the dynamic size option, e.g. `new SimdFFT<float, Transform::Complex>(N)`.

I've tested in a few computers / browsers, and it performs at 45-60% the speed of the current fastest implementation. for example here's a screenshot of the benchmark running in an Apple M3 Pro on Safari:

<img width="907" height="680" alt="Image" src="https://github.com/user-attachments/assets/5d1bffc7-c30a-464b-b2b3-665377b2b419" />

you can [test in your setup](https://mildsunrise.github.io/web-fft-benchmark/)! Remember the speeds above use FFTW's convention (including the 5 factor) and that some browsers greatly restrict CPU time available to a website when it is not focused / visible.

**Comment by RalphSteinhagen** [comment]
@mildsunrise thanks a lot and great results! I'm surprised by the factor-of-two difference, and while the SimdFFT is aiming more towards simpler than fast, I would still love to investigate why there's such a (relatively) significant difference and whether there are any low-hanging fruits I have missed.

Would you have time/interest to iterate on this? Can one reach you via pm? This discussion goes well beyond the FFTw discussion (started here). 😄

**Comment by mildsunrise** [comment]
sure, happy to be messaged, hit me up on email or mastodon :) FWIW, your performance is similar (if not a bit higher) than what I recall from my first SIMD attempt. the source code for the best performing implementation in the chart above (radix 4 packed SIMD) can be found [here](https://github.com/mildsunrise/web-fft-benchmark/blob/main/impls/radix4-simd2-as.ts) (hot part) and [here](https://github.com/mildsunrise/web-fft-benchmark/blob/650d245e7f6537272f7160170991ffcc9712e20a/worker.mjs#L54-L101) (JS) but my code quality is terrible, sorry about that :_ also keep in mind my implementations are only complex, 32-bit float and for powers of 2, while yours is much more general. (I was halfway through making a real FFT but lost motivation in the meantime)

but if I recall correctly, the main trick for that implementation is that the first pass (first depth execution, which can be radix 2 or radix 4 depending on if the FFT size is an odd or even power of 2) also packs the incoming data into [real0, real1, real2, real3, imag0, imag1, imag2, imag3, real4, ...] layout. the rest of the passes operate in this packed layout, except for the last pass, which unpacks the samples back.

**Comment by RalphSteinhagen** [comment]
@mildsunrise great! PM-ed you already. There are other aspects that I'd love your feedback on.

I had a quick chat with Matthias (@mattkretz), and we agreed that the issue is probably not related to the algorithm choice itself (which is somewhat similar/identical), but rather related to Emscripten (i.e. LLVM/clang). The current state of SIMD integration in libc++ is a bit behind. Presently, most of the SIMD in the libc++ is by virtue of the auto-vectorisation that is effectively enabled by using `-msimd128` or `-mrelaxed-simd`, but that is probably not aware of -- notably -- the shuffle/permutation operations that are described in and critical to the FFT algorithm.

However, the good LLVM-people are working on [this](https://github.com/llvm/llvm-project/pull/139919), and this will probably change and catch up soonish (N.B. `std::simd` is now officially part of the C++26 standard).

**Comment by mildsunrise** [comment]
sorry for kinda disappearing, posting here for @RalphSteinhagen and/or anyone who may find it useful: I found an implementation that is substantially faster than mine in many cases, [wasmfft](https://github.com/IQEngine/WebFFT/tree/main/lib/wasmfft). I've added it to the benchmark. sadly neither my implementation(s) nor this one are trivial to use in a C/C++/Rust application, for different reasons in each case (mine does the buffer management / twiddle factor precomputation in JS, so that'd have to be ported to native, and wasmfft uses a compile-time fixed FFT configuration (specified via WASM globals). my implementation still seems to be substantially faster under JSC on M3 Max, see https://github.com/IQEngine/WebFFT/pull/92#issuecomment-4997662587

**Comment by RalphSteinhagen** [comment]
@mildsunrise thanks for the info and I am sorry for the late reply. _I've been travelling without any IDE for a bit more than four weeks. 🌴 🌴 🌞 🐘 🐅 🐒 ⛰️ 🌴 🌴._ 

Our/my primary goal was portability. HW architecture independence and compatibility with C/C++, since we use the same algorithm -- WASM aside -- also on native and embedded targets (CPU|MCU|GPU via GCC, Clang, MSVC, SYCL, ...). Also, code size and single-header only were constraints w.r.t. maintenance since our implementation is needed and to evolve in the long run. Of course, there are faster implementations, but many of these (notably on ARM and MCUs) often hardcode specific SIMD instructions. 

Using `std::simd<T>` allows us to follow the evolution of the compiler vendors who know best which instructions are available, their timings, caching requirements, etc., rather than us handwriting code.
